File: python/day2/day2.py
from typing import List
from collections import defaultdict


def read_file(location: str) -> List[str]:
    with open(location, 'r') as inp_file:
        return [i.strip() for i in inp_file]


def count_tokens(code: str) -> (int, int):
    token_count = defaultdict(lambda: 0)
    for i in code:
        token_count[i] += 1
    values = token_count.values()
    twice_repeat = 1 if 2 in values else 0
    thrice_repeat = 1 if 3 in values else 0
    return twice_repeat, thrice_repeat


def solve_part_one(lines: List[str]):
    two, three = 0, 0
    for l in lines:
        twice, thrice = count_tokens(l)
        two += twice
        three += thrice
    return two * three


def difference(code: str, other: str) -> int:
    if len(code) != len(other):
        return 0
    diff = 0
    for i, _ in enumerate(code):
        if code[i] != other[i]:
            diff += 1
        if diff > 1:
            return diff
    return diff


def common_chars(code: str, other: str) -> str:
    result = ""
    for i, val in enumerate(code):
        if code[i] == other[i]:
            result += val
    return result 


def solve_part_two(lines: List[str]) -> int:
    for first in lines:
        for second in lines[1:]:
            diff = difference(first, second)
            if diff == 1:
                return common_chars(first, second)


if __name__ == '__main__':
    lines = read_file('input.txt')
    print("Part 1 Check code is {}".format(solve_part_one(lines)))
    print("Part 2 Shared Chars is {}".format(solve_part_two(lines)))

File: golang/day2/day2.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io/ioutil"
	"log"
	"strings"
)

func countExists(data map[rune]int, target int) int {
	for _, v := range data {
		if v == target {
			return 1
		}
	}
	return 0
}

func parseLine(target string) (int, int) {
	line := strings.TrimSuffix(target, "\n")
	info := make(map[rune]int)
	for _, t := range line {
		_, exists := info[t]
		if exists {
			info[t]++
		} else {
			info[t] = 1
		}
	}
	return countExists(info, 2), countExists(info, 3)
}

func solvePartOne(input []byte) {
	firstCount, secoundCount := 0, 0
	scanner := bufio.NewScanner(bytes.NewReader(input))
	for scanner.Scan() {
		two, three := parseLine(scanner.Text())
		firstCount += two
		secoundCount += three
	}
	result := firstCount * secoundCount
	fmt.Printf("Part 1: result is %d\n", result)
}

type Record struct {
	Chars []string
	Len   int
}

func difference(s1 []string, s2 []string) int {
	diff := 0
	for i := 0; i < len(s1); i++ {
		if s1[i] != s2[i] {
			diff++
		}
	}
	return diff
}

func sharedChars(s1 []string, s2 []string) string {
	finalString := ""
	for i := 0; i < len(s1); i++ {
		if s1[i] == s2[i] {
			finalString += s1[i]
		}
	}
	return finalString
}

func solvePartTwo(input []byte) {
	all := []Record{}
	scanner := bufio.NewScanner(bytes.NewReader(input))
	for scanner.Scan() {
		line := scanner.Text()
		line = strings.Trim(line, "\n")
		record := Record{Chars: strings.Split(line, ""), Len: len(line)}
		all = append(all, record)
	}
	for _, rcd := range all {
		for _, rcd2 := range all[1:] {
			diff := difference(rcd.Chars, rcd2.Chars)
			//fmt.Println(diff)
			if diff == 1 {
				fmt.Printf("Part two result is: %s\n", sharedChars(rcd.Chars, rcd2.Chars))
				return
			}
		}
	}
}

func main() {
	input, err := ioutil.ReadFile("input.txt")
	if err != nil {
		log.Fatalf("Encountered fatal error, %s", err)
	}
	solvePartOne(input)
	solvePartTwo(input)
}
